### PR TITLE
iproute2: stateless configuration

### LINKED ISSUE #263445: NixOS fails to build: iproute2-6.5.0/etc': No such file or directory
### Describe the bug
I was just trying to upgrade my NixOS installation to current nixos-unstable (2023-10-24) but the build fails with the following error:

```nix
$ nix --extra-experimental-features nix-command flakes build .#nixosConfigurations."optibox".config.system.build.toplevel --verbose --out-link /tmp/nixos-rebuild.jKcMUv/result
error:
       … while calling the 'head' builtin

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/attrsets.nix:850:11:

          849|         || pred here (elemAt values 1) (head values) then
          850|           head values
             |           ^
          851|         else

       … while evaluating the attribute 'value'

         at /nix/store/z1nvpjx9vd4151vx2krxzmx2p1a36pf9-source/lib/modules.nix:807:9:

          806|     in warnDeprecation opt //
          807|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          808|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: getting status of '/nix/store/fijy0dlyk5b06ay7nxg88lr7bp13hpjb-iproute2-6.5.0/etc': No such file or directory
```

Seems like iproute2's `etc` is missing? 

I'm not sure what other data might be helpful. 

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.7, NixOS, 23.11 (Tapir), 23.11.20231019.7c9cc5a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.1`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/4jfc6vrkmq7z5pb651jh5b4kra5f1kwp-source`
```

### PR DESCRIPTION
**Pull request by andresilva**
## Description of changes

Fixes https://github.com/NixOS/nixpkgs/issues/263445

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/config/iproute2.nix [hunk 1]
```diff
@@ -18,15 +18,15 @@ in
...
-    environment.etc."iproute2/ematch_map"  = { mode = "0644"; text = fileContents "${pkgs.iproute2}/etc/iproute2/ematch_map";  };
-    environment.etc."iproute2/group"       = { mode = "0644"; text = fileContents "${pkgs.iproute2}/etc/iproute2/group";       };
-    environment.etc."iproute2/nl_protos"   = { mode = "0644"; text = fileContents "${pkgs.iproute2}/etc/iproute2/nl_protos";   };
-    environment.etc."iproute2/rt_dsfield"  = { mode = "0644"; text = fileContents "${pkgs.iproute2}/etc/iproute2/rt_dsfield";  };
-    environment.etc."iproute2/rt_protos"   = { mode = "0644"; text = fileContents "${pkgs.iproute2}/etc/iproute2/rt_protos";   };
-    environment.etc."iproute2/rt_realms"   = { mode = "0644"; text = fileContents "${pkgs.iproute2}/etc/iproute2/rt_realms";   };
-    environment.etc."iproute2/rt_scopes"   = { mode = "0644"; text = fileContents "${pkgs.iproute2}/etc/iproute2/rt_scopes";   };
-    environment.etc."iproute2/rt_tables"   = { mode = "0644"; text = (fileContents "${pkgs.iproute2}/etc/iproute2/rt_tables")
+    environment.etc."iproute2/bpf_pinning" = { mode = "0644"; text = fileContents "${pkgs.iproute2}/lib/iproute2/bpf_pinning"; };
```
Comment (K900) [inline nixos/modules/config/iproute2.nix:21]: These should probably be removed now: https://github.com/iproute2/iproute2/commit/0a0a8f12fa1b03dd0ccbebf5f85209d1c8a0f580

**Review by K900 — CHANGES_REQUESTED** [review]

**Comment by K900** [comment]
Also, I won't block this fix on that, but it would be nice to have a test for the module added, and linked to the iproute2 package.

File: nixos/modules/config/iproute2.nix (hunk 1 above)
Comment (andresilva) [reply nixos/modules/config/iproute2.nix:21]: Not sure if removing all these files is correct, at least `rt_tables` if you grep for this in nixpkgs there are multiple VPN packages expecting this file to exist and claiming to write to it.

**Comment by andresilva** [comment]
I have to admit that I have no idea about how iproute2 works, I just made this PR in an effort to make my system build. So anything more involved than what's being done here should really be done by someone that actually knows what they are doing.

File: nixos/modules/config/iproute2.nix (hunk 1 above)
Comment (K900) [reply nixos/modules/config/iproute2.nix:21]: `rt_tables` is also customized by the module, so it will have to stay. The rest should be removed though.

**Comment by K900** [comment]
You should be able to just disable the option for now.

**Comment by andresilva** [comment]
I updated the PR to remove all the files except `rt_tables`.

> You should be able to just disable the option for now.

It's enabled indirectly by a VPN that I use. Feel free to close the PR if you think a more appropriate solution is in-order.

**Comment by NickCao** [comment]
See https://github.com/NixOS/nixpkgs/pull/263472 for a more appropriate fix, (and somewhat related discussions in https://github.com/NixOS/nixpkgs/pull/242453)

**Comment by naturallaw777** [comment]
I too am getting the same issue.

**Comment by Lillecarl** [comment]
@K900 Why should they be removed since the stateless configuration change? Are we expected to do these copies ourselves or what's up? :)

**Comment by K900** [comment]
No, there's just no point in having those files there anymore.

**Comment by andresilva** [comment]
> @K900 Why should they be removed since the stateless configuration change? Are we expected to do these copies ourselves or what's up? :)

From what I understood if the files aren't there (in /etc/) iproute2 will fallback to using the package files from /lib/. So there's no point in copying them except for `rt_tables`.

**Review by purcell — APPROVED** [review]
I can confirm this resolves #263445 for me (on aarch64-linux), fwiw.

**Review by K900 — APPROVED** [review]

**Pull request merged**